chromium won't open fully maximized on launch

I just updated to the latest labwc-git build and chrome is now throwing this error when launching, it wasn't happening with wlroots 0.17.

```
00:01:55.849 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae0d0f0
00:01:55.849 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae0d0f0
00:01:57.989 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eac47720
00:01:57.989 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eac47720
00:02:07.377 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae1e6a0
00:02:07.377 [ERROR] [types/xdg_shell/wlr_xdg_surface.c:169] A configure is scheduled for an uninitialized xdg_surface 0x5a15eae1e6a0
```

I noticed this because chromium won't open fully maximized on launch (a small 1/2 inch gap is left at the bottom between the display's border and chromium's border), it is only after a few milliseconds that it will restore the window to its full maximized state.

I don't know who is at fault here, if it's either chromium, labwc or wlroots... but I figured I could start reporting the issue here. THis does not happen with other programs only with chromium.

> now I only get the "ignoring maximize before map" error.
> 
> ```
> 00:00:02.250 [ERROR] [../src/xdg.c:288] ignoring maximize before map
> 00:00:04.372 [ERROR] [../src/xdg.c:288] ignoring maximize before map
> ```

Yeah, those are expected, they are just for debugging. I'll remove those from the PR before we merge it.

I see similar a similar message with `fuzzel` (wlr-layer-surface client):

```
00:00:02.971 [ERROR] [types/wlr_layer_shell_v1.c:296] A configure is sent to an uninitialized wlr_layer_surface_v1 0x604b6c77e060
```

I haven't investigated the cause yet.

--- a/src/xdg.c
+++ b/src/xdg.c
@@ -651,10 +651,6 @@ xdg_toplevel_view_map(struct view *view)
 	if (!view->been_mapped) {
 		init_foreign_toplevel(view);
 
-		/*
-		 * FIXME: is this needed or is the earlier logic in
-		 * xdg_surface_new() enough?
-		 */
 		if (view_wants_decorations(view)) {
 			view_set_ssd_mode(view, LAB_SSD_MODE_FULL);
 		} else {
@@ -907,13 +903,6 @@ xdg_toplevel_new(struct wl_listener *listener, void *data)
 	CONNECT_SIGNAL(xdg_surface, xdg_toplevel_view, new_popup);
 
 	wl_list_insert(&server->views, &view->link);
-
-	/* FIXME: is view_wants_decorations() reliable this early? */
-	if (view_wants_decorations(view)) {
-		view_set_ssd_mode(view, LAB_SSD_MODE_FULL);
-	} else {
-		view_set_ssd_mode(view, LAB_SSD_MODE_NONE);
-	}
 }
 
 void

--- a/src/xdg.c
+++ b/src/xdg.c
@@ -76,6 +76,18 @@ handle_new_popup(struct wl_listener *listener, void *data)
 }
 
 static void
+set_fullscreen_from_request(struct view *view,
+		struct wlr_xdg_toplevel_requested *requested)
+{
+	if (!view->fullscreen && requested->fullscreen
+			&& requested->fullscreen_output) {
+		view_set_output(view, output_from_wlr_output(view->server,
+			requested->fullscreen_output));
+	}
+	view_set_fullscreen(view, requested->fullscreen);
+}
+
+static void
 do_late_positioning(struct view *view)
 {
 	struct server *server = view->server;
@@ -100,11 +112,28 @@ handle_commit(struct wl_listener *listener, void *data)
 {
 	struct view *view = wl_container_of(listener, view, commit);
 	struct wlr_xdg_surface *xdg_surface = xdg_surface_from_view(view);
+	struct wlr_xdg_toplevel *toplevel = xdg_toplevel_from_view(view);
 	assert(view->surface);
 
 	if (xdg_surface->initial_commit) {
 		wlr_log(WLR_DEBUG, "scheduling configure");
 		wlr_xdg_surface_schedule_configure(xdg_surface);
+		/*
+		 * Handle initial fullscreen/maximize requests immediately after
+		 * scheduling the initial configure event (before it is sent) in
+		 * order to send the correct size and avoid flicker.
+		 *
+		 * In normal (non-fullscreen/maximized) cases, the initial
+		 * configure event is sent with a zero size, which requests the
+		 * application to choose its own size.
+		 */
+		if (toplevel->requested.fullscreen) {
+			set_fullscreen_from_request(view, &toplevel->requested);
+		}
+		if (toplevel->requested.maximized) {
+			view_maximize(view, VIEW_AXIS_BOTH,
+				/*store_natural_geometry*/ true);
+		}
 		return;
 	}
 
@@ -311,30 +340,38 @@ static void
 handle_request_maximize(struct wl_listener *listener, void *data)
 {
 	struct view *view = wl_container_of(listener, view, request_maximize);
+	struct wlr_xdg_toplevel *toplevel = xdg_toplevel_from_view(view);
+
+	if (!toplevel->base->initialized) {
+		/*
+		 * Do nothing if we have not received the initial commit yet.
+		 * We will maximize the view in the commit handler.
+		 */
+		return;
+	}
+
 	if (!view->mapped && !view->output) {
 		view_set_output(view, output_nearest_to_cursor(view->server));
 	}
-	bool maximized = xdg_toplevel_from_view(view)->requested.maximized;
+	bool maximized = toplevel->requested.maximized;
 	view_maximize(view, maximized ? VIEW_AXIS_BOTH : VIEW_AXIS_NONE,
 		/*store_natural_geometry*/ true);
-}
-
-static void
-set_fullscreen_from_request(struct view *view,
-		struct wlr_xdg_toplevel_requested *requested)
-{
-	if (!view->fullscreen && requested->fullscreen
-			&& requested->fullscreen_output) {
-		view_set_output(view, output_from_wlr_output(view->server,
-			requested->fullscreen_output));
-	}
-	view_set_fullscreen(view, requested->fullscreen);
 }
 
 static void
 handle_request_fullscreen(struct wl_listener *listener, void *data)
 {
 	struct view *view = wl_container_of(listener, view, request_fullscreen);
+	struct wlr_xdg_toplevel *toplevel = xdg_toplevel_from_view(view);
+
+	if (!toplevel->base->initialized) {
+		/*
+		 * Do nothing if we have not received the initial commit yet.
+		 * We will fullscreen the view in the commit handler.
+		 */
+		return;
+	}
+
 	if (!view->mapped && !view->output) {
 		view_set_output(view, output_nearest_to_cursor(view->server));
 	}
@@ -876,24 +913,6 @@ xdg_toplevel_new(struct wl_listener *listener, void *data)
 		view_set_ssd_mode(view, LAB_SSD_MODE_FULL);
 	} else {
 		view_set_ssd_mode(view, LAB_SSD_MODE_NONE);
-	}
-
-	/*
-	 * Handle initial fullscreen/maximize requests. This needs to be
-	 * done early (before map) in order to send the correct size to
-	 * the initial configure event and avoid flicker.
-	 *
-	 * Note that at this point, wlroots has already scheduled (but
-	 * not yet sent) the initial configure event with a size of 0x0.
-	 * In normal (non-fullscreen/maximized) cases, the zero size
-	 * requests the application to choose its own size.
-	 */
-	if (toplevel->requested.fullscreen) {
-		set_fullscreen_from_request(view, &toplevel->requested);
-	}
-	if (toplevel->requested.maximized) {
-		view_maximize(view, VIEW_AXIS_BOTH,
-			/*store_natural_geometry*/ true);
 	}
 }
 

